src/Equinox.MessageDb/MessageDb.fs

```diff
@@ -0,0 +1,435 @@
+    /// Unless <c>LoadOption.AllowStale</c> is used, each cache hit still incurs a roundtrip to load any subsequently-added events.
+    | FixedTimeSpan of ICache * period : TimeSpan
+    /// Prefix is used to segregate multiple folds per stream when they are stored in the cache.
+    /// Semantics are identical to <c>SlidingWindow</c>.
+    | SlidingWindowPrefixed of ICache * window : TimeSpan * prefix : string
+
+type MessageDbCategory<'event, 'state, 'context>(resolveInner, empty) =
+    inherit Equinox.Category<'event, 'state, 'context>(resolveInner, empty)
+    new (   context : MessageDbContext, codec : IEventCodec<_, _, 'context>, fold, initial,
+            [<O; D(null)>]?caching,
+            [<O; D(null)>]?access) =
+        let inner = Category<'event, 'state, 'context>(context, codec, ?access = access)
+        let readCacheOption =
+            match caching with
+            | None -> None
+            | Some (CachingStrategy.SlidingWindow (cache, _))
+            | Some (CachingStrategy.FixedTimeSpan (cache, _)) -> Some (cache, null)
+            | Some (CachingStrategy.SlidingWindowPrefixed (cache, _, prefix)) -> Some (cache, prefix)
+        let folder = Folder<'event, 'state, 'context>(inner, fold, initial, ?readCache = readCacheOption)
+        let category : ICategory<_, _, 'context> =
+            match caching with
+            | None -> folder :> _
+            | Some (CachingStrategy.SlidingWindow (cache, window)) ->
+                Caching.applyCacheUpdatesWithSlidingExpiration cache null window folder Token.supersedes
+            | Some (CachingStrategy.FixedTimeSpan (cache, period)) ->
+                Caching.applyCacheUpdatesWithFixedTimeSpan cache null period folder Token.supersedes
+            | Some (CachingStrategy.SlidingWindowPrefixed (cache, window, prefix)) ->
+                Caching.applyCacheUpdatesWithSlidingExpiration cache prefix window folder Token.supersedes
+        let resolveInner streamIds = struct (category, StreamName.Internal.ofCategoryAndStreamId streamIds, ValueNone)
+        let empty = struct (context.TokenEmpty, initial)
+        MessageDbCategory(resolveInner, empty)
+
+
+type MessageDbConnector(
+    connectionString : string, ?readOnlyConnectionString : string,
+    [<O; D(null)>]?readRetryPolicy, [<O; D(null)>]?writeRetryPolicy) =
+        member _.Connect() =
+            let connectToDb connectionString ct = task {
+                let conn = new NpgsqlConnection(connectionString)
+                do! conn.OpenAsync(ct)
+                return conn }
+            let writeConnection = connectToDb connectionString
+            let readConnection =
+                match readOnlyConnectionString with
+                | Some connectionString -> connectToDb connectionString
+                | None -> writeConnection
+            MessageDbClient(writeConnection), MessageDbClient(readConnection)
+
+        member x.Establish() : MessageDbConnection =
```

Comment:
I think the connection and the client should probably be merged
- assuming a fresh conn is needed per call, the client should take 2x `unit -> npgclient` 

(the existing layout/config is based on how it needs to work for esdb - the conn holds 2 clients and then the call uses the right one with the right retry polcy depending on whether its reading or writing but critically when resyncing - in that case, the reads need to be via the write connection (to the leader node). Because that complicaiton of reading via the write connection does not arise, the client just needs to select the right connection(string) based on whether it's a read or a write. That does leave a puzzle as to where the retry policy goes though...)

The other direction to go is to make the client methods stateless and pass them a `CT -> npgsqlconn` or similar ?

Feel free to leave it, and I can do it as a follow-up as I'm not entirely sure how it will/should land exactly work - it shouldnt take me long as I've juggled these about quite a bit in the different clients!

Reply:
> Feel free to leave it, and I can do it as a follow-up 

I'm happy to make some more changes here, I agree this feels like an unnecessary type to have.

I'm also aware that Npgsql 7 is landing soon and will upend whatever we do here, so I'm not to hung up on any choice made 😄

Reply:
> when resyncing - in that case, the reads need to be via the write connection

I would assume a similar requirement for a sql based store where there's a main/read-replica relationship. When resyncing you definitely want to read from the main as the replica can be behind.

Reply:
>I'm also aware that Npgsql 7 is landing 

Man am I glad the world has not triggered me into understanding anything in that world to date! It does feel the way you have the npgsql wiring in another file with `task {` is a good way to split the responsibilities in terms of localising any impact as and when an update gets triggered

Reply:
>I would assume a similar requirement for a sql based store where there's a main/read-replica relationship. When resyncing you definitely want to read from the main as the replica can be behind.

Hm, that's a very good point (but I think that's better addressed by having a `forceReadFromLeader : bool` passed into the client rather than having the switching/selection of the connection be sprinkled all over the `module Read` like ti started in the esdb impl would be a good cleanup if at all possible (I am willing to bet the fact this is going on is far from obvious to the average person scanning it))

Reply:
Wow that was quick and neat; feels right!